Derive rocksdb_fifo_shred_storage_size based on limit-ledger-size

yhchiang-sol opened the pull request:
### Problem
When FIFO compaction is used while --rocksdb_fifo_shred_storage_size
is unspecified, the FIFO shred storage size is set to a const default based
on the default `--limit-ledger-size`.

### Summary of the Change
When --rocksdb_fifo_shred_storage_size is unspecified, it is now
derived from `--limit-ledger-size` by reserving 1500 bytes for each
shred.

steviez commented on validator/src/main.rs:
```diff
@@ -392,6 +392,18 @@ fn hash_validator(hash: String) -> Result<(), String> {
+    // can make sure FIFO rarely evits data/coding shreds that is younger
+    // than the configured max_ledger_shreds while avoiding allocating too
+    // much disk space for data/coding shreds.
+    const AVG_SHRED_SIZE_BYTE: u64 = 3200;
```
This number assumes 1:1 data:coding, right? A little more detail on where this number comes from would probably be good

yhchiang-sol replied:
This is the total shred size, including both data and coding shred.

steviez replied:
I meant to include more detail on where you got 3200. Ie, shreds are 1228 bytes, we have 1:1 data and coding so double that, and add some margin for meta-data to get to 3200.

Or did you just double this number from ledger_cleanup_service comments

yhchiang-sol replied:
I think the confusing part is the `/2` in the final computation.  

    Some(max_ledger_shreds) => (max_ledger_shreds / 2) * AVG_SHRED_SIZE_BYTE
    
I will change it to 1600 instead and remove the `/2` and add more comments.

steviez commented on validator/src/main.rs:
```diff
@@ -1052,14 +1063,14 @@ pub fn main() {
-                       The suggested value is 50% of your ledger storage size in bytes."),
+                       The suggested value is 50% of your ledger storage size. \
+                       If this argument is unspecified, we will assign a proper \
+                       value based on --limit-shred-size."),
```
based on `--limit-ledger-size`

steviez commented on validator/src/main.rs:
```diff
@@ -392,6 +392,20 @@ fn hash_validator(hash: String) -> Result<(), String> {
+    // than the configured max_ledger_shreds while avoiding allocating too
+    // much disk space for data/coding shreds.
+    const AVG_SHRED_SIZE_BYTE: u64 = 3200;
+    match vc.max_ledger_shreds {
+        Some(max_ledger_shreds) => (max_ledger_shreds / 2) * AVG_SHRED_SIZE_BYTE,
+        None => DEFAULT_ROCKS_FIFO_SHRED_STORAGE_SIZE_BYTES,
```
Given how validator is written, `vc.max_ledger_shreds` should always be a `Some(...)`; wondering if we should rework this.

Also wondering if `DEFAULT_ROCKS_FIFO_SHRED_STORAGE_SIZE_BYTES` should match the size that would result from the estimate here with `DEFAULT_MAX_LEDGER_SHREDS`

yhchiang-sol replied:
On top of this PR, the DEFAULT_ROCKS_FIFO_SHRED_STORAGE_SIZE_BYTES will be removed in https://github.com/solana-labs/solana/pull/27965.  I've thinking about merging these two PRs, but I feel it's probably better to have one PR for one purpose: this PR focuses on making FIFO arguments public, and the other PR focuses on deprecate unused default.

steviez replied:
Ok, fair enough about the other PR addressing the default.

steviez commented on validator/src/main.rs:
```diff
@@ -392,6 +392,20 @@ fn hash_validator(hash: String) -> Result<(), String> {
+    // much disk space for data/coding shreds.
+    const AVG_SHRED_SIZE_BYTE: u64 = 3200;
+    match vc.max_ledger_shreds {
+        Some(max_ledger_shreds) => (max_ledger_shreds / 2) * AVG_SHRED_SIZE_BYTE,
```
This is a reduction in how many shreds we'll store right? If max_ledger_shreds == 200,000,000 (the default), then:
```
( 200,000,000 / 2 ) * 3200 bytes = 320,000,000,000 bytes = 320 GB
```
That is 320 GB total, so 160 GB per column family. Or, essentially, 100 million data and 100 million coding whereas the current configuration is setup to take 200 million data shreds

yhchiang-sol replied:
The way I am thinking this is that: given the default 200,000,000 shreds require the suggested 500GB.  We previously observed that data + coding shreds occupy ~50% of the ledger data (no rpc case).  This gives us ~250GB.

This PR increases to 320GB to give an extra storage buffer to FIFO to prevent the case where FIFO cleans up data earlier than our ledger cleanup service.

steviez replied:
The updated code snippet:
```rust
    const RESERVED_BYTES_PER_SHRED: u64 = 1600;
    match vc.max_ledger_shreds {
        Some(max_ledger_shreds) => max_ledger_shreds * RESERVED_BYTES_PER_SHRED,
```
With defaults limit-ledger-size value (200 million):
```
200,000,000 shreds * 1600 bytes / shred = 320,000,000,00 = 320 GB
```
That output of above function is consumed here:
```rust
                "fifo" => match matches.value_of("rocksdb_fifo_shred_storage_size") {
                    None => ShredStorageType::rocks_fifo(default_fifo_shred_storage_size(
                        &validator_config,
                    )),
```
`ShredStorageType::rocks_fifo()` implementation:
```rust
impl ShredStorageType {
    /// Returns ShredStorageType::RocksFifo where the specified
    /// `shred_storage_size` is equally allocated to shred_data_cf_size
    /// and shred_code_cf_size.
    pub fn rocks_fifo(shred_storage_size: u64) -> ShredStorageType {
        ShredStorageType::RocksFifo(BlockstoreRocksFifoOptions::new(shred_storage_size))
    }
```
`BlockstoreRocksFifoOptions::new()` implementation:
```rust
impl BlockstoreRocksFifoOptions {
    fn new(shred_storage_size: u64) -> Self {
        Self {
            shred_data_cf_size: shred_storage_size / 2,
            shred_code_cf_size: shred_storage_size / 2,
        }
    }
}
```

That 320 GB is getting divided up at the bottom here, so 160 GB per column right?

yhchiang-sol replied:
> That 320 GB is getting divided up at the bottom here, so 160 GB per column right?

Yes, it is 160GB per column.

yhchiang-sol replied:
Oh, oh, oh!  I was indeed wrong.  It is ~50% if we consider RPC calls, and nearly 95% if we don't consider RPC calls.

So it would be 320GB per column if we want to add some extra buffer.  Maybe 300GB each is already enough.  I will update the PR and the description of the argument to include this information.

steviez commented on validator/src/main.rs:
```diff
@@ -392,6 +392,20 @@ fn hash_validator(hash: String) -> Result<(), String> {
+/// Returns the default shred storage size (include both data and coding
+/// shreds) based on the validator config.
+fn default_fifo_shred_storage_size(vc: &ValidatorConfig) -> u64 {
+    // The average shred size is around 1228 bytes.
```
"average" ==> "max"

steviez commented on validator/src/main.rs:
```diff
@@ -392,6 +392,20 @@ fn hash_validator(hash: String) -> Result<(), String> {
+    const RESERVED_BYTES_PER_SHRED: u64 = 1600;
+    match vc.max_ledger_shreds {
+        Some(max_ledger_shreds) => max_ledger_shreds * RESERVED_BYTES_PER_SHRED,
+        None => DEFAULT_ROCKS_FIFO_SHRED_STORAGE_SIZE_BYTES,
```
I misspoke in another comment; `max_ledger_shreds` will not always be a `Some(...)`. However, when it is a `None`, we should not enforce any limit here and the ledger should be allowed to grow indefinitely.

A usecase for this is our warehouse nodes that collect an epoch of ledger data without purging. I'm not sure we would switch over the warehouse nodes to fifo; but regardless, we shouldn't allow for possibility of unintended behavior

yhchiang-sol replied:
Interesting.  Can I know what is the current settings of our warehouse nodes?  Does their LedgerCleanupService also purge nothing because `--limit-ledger-size` is not set?

If the answer is yes, then warehouse nodes are already configured to grow indefinitely.  In that case, they are also safe to switch to FIFO if we allow the ledger to grow indefinitely when max_ledger_shreds is None.

steviez replied:
https://github.com/solana-labs/solana/blob/cdb95a362faf66e9384ece57054b45e805d0caa2/core/src/tvu.rs#L297-L306

yhchiang-sol replied:
Got it!  So it should be unlimited for FIFO in that case.  Let me update the PR.

steviez commented:
Making these arguments public implies we support them - something else to consider, should we have `LedgerCleanupService` running on shred columns if we're in FIFO mode ?

yhchiang-sol commented:
> Making these arguments public implies they are good to go - something else to consider, should we have LedgerCleanupService running on shred columns if we're using fifo. Ie, anything bad about the purge calls ?

LedgerCleanupService should keep running on FIFO data and coding shreds as they also issue range_delete to make data consistent across different column families.  Note that for the actual clean-up, we rely on RocksDB's background compaction to do the job.  In that case, it could be FIFO or level compaction, depending on the setting.

One thing in my mind is that when the replay progress is seriously behind that causes some shreds older than the `--limit-ledger-size` are still not replayed.  In this case, our LedgerCleanupService will wait for the replay (and risks running out of disk space), while FIFO will not, and this is the reason why I am giving FIFO an extra storage buffer.  A better way to handle this is to throttle the validator as the input speed is faster than the replay speed somehow.

steviez commented:
> One thing in my mind is that when the replay progress is seriously behind that causes some shreds older than the `--limit-ledger-size` are still not replayed. In this case, our LedgerCleanupService will wait for the replay (and risks running out of disk space), while FIFO will not, and this is the reason why I am giving FIFO an extra storage buffer. A better way to handle this is to throttle the validator as the input speed is faster than the replay speed somehow.

Is there an easy way to know if data that fifo evicts was greater than lowest_cleanup_slot? I'm guessing not, but if we had the max key of file that was deleted we'd know

yhchiang-sol commented:
> Is there an easy way to know if data that fifo evicts was greater than lowest_cleanup_slot? I'm guessing not, but if we had the max key of file that was deleted we'd know

EventListener API can do the trick, but I think it's not yet available in rust.

Another way is to keep monitoring the oldest sst file as well as the configured column family size.  If the size is close to the configured FIFO limit and the youngest key of the oldest file is not older than the lowest_cleanup_slot, then we know this file is likely to be evicted too early.

yhchiang-sol commented:
Addressed comments.  This PR now takes a step back and only makes the default FIFO shred storage size smarter.

> Another way is to keep monitoring the oldest sst file as well as the configured column family size. If the size is close to the configured FIFO limit and the youngest key of the oldest file is not older than the lowest_cleanup_slot, then we know this file is likely to be evicted too early.

Luckily to find that CompactionOptionsFIFO::max_table_files_size is actually dynamically changeable!  So we can do this smarter by dynamically changing CompactionOptionsFIFO::max_table_files_size in LedgerCleanupService.  It has three cases:

* If the latest key in the oldest sst file is still newer than the current purge_slot (i.e., the oldest sst file is not eligible to purge), then LedgerCleanupService temporarily set CompactionOptionsFIFO::max_table_files_size to u64::MAX.

* If the oldest file is eligible to purge but the total CF size is greater than the configured threshold, then we set CompactionOptionsFIFO::max_table_files_size to `CF_size - 1` to allow FIFO to purge the oldest file.

* If the oldest file is eligible to purge but the total CF size is smaller than the configured threshold, then we set
CompactionOptionsFIFO::max_table_files_size to the configured threshold.

steviez commented on validator/src/main.rs:
```diff
@@ -1057,9 +1070,10 @@ pub fn main() {
                 .help("The shred storage size in bytes. \
-                       The suggested value is 50% of your ledger storage size in bytes."),
+                       The suggested value is at least 50% of your ledger storage size. \
+                       If this argument is unspecified, we will assign a proper \
+                       value based on --limit-ledger-size."),
```
Now that we discovered we had previously not been honoring the case where `--limit-ledger-size` was unspecified (to allow unbounded growth), should we tweak the help a little to explain what we'll do in that case? I don't love my wording here, but to illustrate the point I'm making:

> If this argument is unspecified, we will choose a proper value consistent with --limit-ledger-size value, or allow unbounded storage size growth if unspecified

yhchiang-sol replied:
> Now that we discovered we had previously not been honoring the case where --limit-ledger-size was unspecified (to allow unbounded growth), should we tweak the help a little to explain what we'll do in that case?

Sure thing.  Btw, to clarify a little bit.  So we have three cases:

* `--limit-ledger-size` is presented and a value is specified, in that case we will use the specified value.
* `--limit-ledger-size` is presented but no value is specified, in this case `DEFAULT_MAX_LEDGER_SHREDS` will be used, which is 200_000_000.
* `--limit-ledger-size` is not presented.  In this case, the ledger size is unlimited by definition and thus is unbounded.

steviez reviewed (approved):
Thanks for sticking with this one - lots of conversation for not a lot of code change, but I think it made us both dig and uncover some corner cases.

One other thing I just thought of - should `rocksdb_fifo_shred_storage_size` arg have an argument function like `requires()` based on `rocksdb_shred_compaction`? Basically, so we could warn a user when they are using `level` but also specified `rocksdb_fifo_shred_storage_size`. Definitely not a use case, but helpful to someone incase they forget to specify the type arg.

I guess since the other arg chooses between `level` or `fifo`, we can't simply do that verification based on whether the flag exists, but would have to do it based on what the value is after we have all the args. Not something that needs to be addressed in this PR, but it just crossed my mind
